dbms/src/Interpreters/DDLWorker.cpp

```diff
@@ -0,0 +1,710 @@
+#include <Interpreters/DDLWorker.h>
+
+#include <Parsers/ASTAlterQuery.h>
+#include <Parsers/ASTQueryWithOnCluster.h>
+#include <Parsers/ParserQuery.h>
+#include <Parsers/parseQuery.h>
+#include <Parsers/queryToString.h>
+
+#include <IO/WriteHelpers.h>
+#include <IO/ReadHelpers.h>
+#include <IO/Operators.h>
+#include <IO/ReadBufferFromString.h>
+
+#include <Storages/IStorage.h>
+#include <DataStreams/OneBlockInputStream.h>
+
+#include <Interpreters/executeQuery.h>
+#include <Interpreters/Cluster.h>
+
+#include <Common/getFQDNOrHostName.h>
+#include <Common/setThreadName.h>
+#include <Common/Stopwatch.h>
+
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeArray.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnArray.h>
+
+#include <zkutil/ZooKeeper.h>
+#include <zkutil/Lock.h>
+#include <Poco/Timestamp.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_ELEMENT_IN_CONFIG;
+    extern const int INVALID_CONFIG_PARAMETER;
+    extern const int UNKNOWN_FORMAT_VERSION;
+    extern const int INCONSISTENT_TABLE_ACCROSS_SHARDS;
+    extern const int INCONSISTENT_CLUSTER_DEFINITION;
+    extern const int TIMEOUT_EXCEEDED;
+    extern const int UNFINISHED;
+}
+
+
+struct DDLLogEntry
+{
+    String query;
+    Strings hosts;
+    String initiator;
+
+    static constexpr char CURRENT_VERSION = '1';
+
+    String toString()
+    {
+        String res;
+        {
+            WriteBufferFromString wb(res);
+
+            writeChar(CURRENT_VERSION, wb);
+            wb << "\n";
+            wb << "query: " << query << "\n";
+            wb << "hosts: " << hosts << "\n";
+            wb << "initiator: " << initiator << "\n";
+        }
+
+        return res;
+    }
+
+    void parse(const String & data)
+    {
+        ReadBufferFromString rb(data);
+
+        char version;
+        readChar(version, rb);
```

Comment:
If version will become 10, there could be some difficulties.
Why don't read/write as integer?

PS. For convenience, I prefer `version: 1`, not just `1`.